File: src/Shortest Path in 1-2 Graph/Code/solution.cpp
class Solution
{
public:
    int shortestPath(int V, int src, int dest, vector<vector<int>> &edges)
    {
        // Total number of nodes after adding virtual nodes
        int totalNodes = V;

        // Adjacency list for the transformed graph
        vector<vector<int>> adj(V + edges.size());

        // Build the transformed graph
        for (auto &edge : edges)
        {
            int u = edge[0];
            int v = edge[1];
            int w = edge[2];

            if (w == 1)
            {
                // Weight 1 edge remains unchanged
                adj[u].push_back(v);
                adj[v].push_back(u);
            }
            else
            {
                // Create one virtual node for a weight 2 edge
                int newNode = totalNodes++;

                // Replace weight 2 with two weight 1 edges
                adj[u].push_back(newNode);
                adj[newNode].push_back(u);

                adj[newNode].push_back(v);
                adj[v].push_back(newNode);
            }
        }

        // Distance array (-1 means unvisited)
        vector<int> dist(totalNodes, -1);

        // Normal BFS queue
        queue<int> q;

        // Start from source
        q.push(src);
        dist[src] = 0;

        while (!q.empty())
        {
            int node = q.front();
            q.pop();

            // Visit every neighbor
            for (int next : adj[node])
            {
                if (dist[next] == -1)
                {
                    dist[next] = dist[node] + 1;
                    q.push(next);
                }
            }
        }

        // Return shortest distance to destination
        return dist[dest];
    }
};

File: src/Shortest Path in 1-2 Graph/Code/solution.js
/**
 * @param {number} V
 * @param {number} src
 * @param {number} dest
 * @param {number[][]} edges
 * @return {number}
 */

class Solution {
  shortestPath(V, src, dest, edges) {
    // Total nodes after adding virtual nodes
    let totalNodes = V;

    // Adjacency list
    const adj = Array.from({ length: V + edges.length }, () => []);

    // Build transformed graph
    for (const [u, v, w] of edges) {
      if (w === 1) {
        // Normal edge
        adj[u].push(v);
        adj[v].push(u);
      } else {
        // Create virtual node
        const newNode = totalNodes++;

        // Replace weight 2 edge
        adj[u].push(newNode);
        adj[newNode].push(u);

        adj[newNode].push(v);
        adj[v].push(newNode);
      }
    }

    // Distance array
    const dist = new Array(totalNodes).fill(-1);

    // BFS queue
    const queue = [];
    let front = 0;

    queue.push(src);
    dist[src] = 0;

    while (front < queue.length) {
      const node = queue[front++];

      for (const next of adj[node]) {
        if (dist[next] === -1) {
          dist[next] = dist[node] + 1;
          queue.push(next);
        }
      }
    }

    // Return shortest distance
    return dist[dest];
  }
}
